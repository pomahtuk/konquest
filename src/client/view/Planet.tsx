import React, { ReactElement } from "react";
import { shallowEqual, useSelector, useDispatch } from "react-redux";
import { Tooltip } from "react-tippy";

import { setDestinationPlanet, setOriginPlanet } from "../actions/game.actions";

import Planet from "../../logic/Planet";

import "react-tippy/dist/tippy.css";
import planetSelectorFunction, { PlanetStoreSlice } from "../selectors/planet.selector";
import RandomImage from "./RandomImage";

export interface PlanetProps {
  planet: Planet;
}

const PlanetElement = ({ planet }: PlanetProps): ReactElement => {
  const { originPlanet, activePlayer, destinationPlanet, currentShipsModifier }: PlanetStoreSlice = useSelector(planetSelectorFunction, shallowEqual);
  const dispatch = useDispatch();

  if (!planet) {
    return <span>Â·</span>;
  }

  const onPlanetSelect = (): void => {
    if (!originPlanet && (!planet.owner || planet.owner.id !== activePlayer.id)) {
      // origin should belong to player
      return;
    } else if (!originPlanet && planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
    } else if (originPlanet && !destinationPlanet) {
      dispatch(setDestinationPlanet(planet));
    } else if (planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
      dispatch(setDestinationPlanet(undefined));
    }
  };

  return (
    <Tooltip
      position="top"
      animation="fade"
      arrow="true"
      animateFill="false"
      html={
        <div>
          <div>Owner: {planet.owner ? planet.owner.screenName : "none"}</div>
          <div>Production: {planet.production}</div>
          <div>Kill percent: {planet.killPercent}</div>
          <div>Ships: {planet.ships - (currentShipsModifier[planet.name] || 0)}</div>
        </div>
      }
      style={{
        position: "absolute",
        top: planet.coordinates.x * 60 - 10,
        left: planet.coordinates.y * 60 - 10,
        width: 60,
        height: 60
      }}
    >
      <div
        style={{
          cursor: "pointer",
          position: "relative",
          top: 10,
          left: 10,
          width: 40,
          height: 40
        }}
        onClick={onPlanetSelect}
      >
        <RandomImage imageKey={planet.name}></RandomImage>
        <span
          style={{
            position: "absolute",
            bottom: 0,
            right: 0,
            zIndex: 999
          }}
        >
          {planet.name}
        </span>
        <br />
      </div>
    </Tooltip>
  );
};

export default PlanetElement;
