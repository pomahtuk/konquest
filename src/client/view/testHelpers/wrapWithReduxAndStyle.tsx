import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
