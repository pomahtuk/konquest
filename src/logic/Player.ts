export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
