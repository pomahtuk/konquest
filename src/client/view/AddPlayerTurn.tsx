import React, { ReactElement, ChangeEvent } from "react";
import Planet from "../../logic/Planet";
import Player from "../../logic/Player";

export interface AddPlayerTurnProps {
  activePlayer: Player;
  originPlanet?: Planet;
  destinationPlanet?: Planet;
  amount: number;
  onOrderAmountChange: (event: ChangeEvent<HTMLInputElement>) => void;
  onAddOrder: () => void;
}

const AddPlayerTurn = ({
  activePlayer,
  amount,
  originPlanet,
  destinationPlanet,
  onOrderAmountChange,
  onAddOrder
}: AddPlayerTurnProps): ReactElement => {
  return (
    <div>
      <h2>{activePlayer.screenName} turn</h2>
      {!originPlanet && <div>Please select origin planet</div>}
      {originPlanet && !destinationPlanet && <div>Please select destination planet</div>}
      {originPlanet && destinationPlanet && (
        <div>
          Sending fleet from {originPlanet.name} to {destinationPlanet.name}. Select amount:
          <input type="number" value={amount} onChange={onOrderAmountChange} />
          <button onClick={onAddOrder}>Add order</button>
        </div>
      )}
    </div>
  );
};

export default AddPlayerTurn;
