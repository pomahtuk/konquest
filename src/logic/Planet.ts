import Player from "./Player";

export interface PlanetCoordinates {
  x: number;
  y: number;
}

export default class Planet {
  public name: string;
  public owner: number | null = null; // reference co playerId
  public ships: number;
  public shipsDue: number | 0 = 0; // only if there are ships en route
  public production: number; // only happens when planet is taken, next turn
  public killPercent: number;
  // Kill percent is a measure of the effectiveness of the ships produced at that planet.
  // Attack fleets take the kill percentage of their planet of departure,
  // and defense fleets use the kill percentage of the planet they are defending.
  public coordinates: PlanetCoordinates;

  public constructor(name: string, player?: Player | null, coordinates: PlanetCoordinates = { x: 0, y: 0 }) {
    // initialize
    this.name = name;
    this.coordinates = coordinates;

    let production = 10;
    const killPercent = 0.5;
    if (!player) {
      production = 8 + Math.floor(Math.random() * 8);
      // TODO: generate killPercent
      this.ships = 0;
    } else {
      this.owner = player.id;
      this.ships = production;
    }

    this.production = production;
    this.killPercent = killPercent;
  }
}

export interface PlanetMap {
  [key: string]: Planet;
}
