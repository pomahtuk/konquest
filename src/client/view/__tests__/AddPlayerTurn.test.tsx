import React from "react";
import { render } from "@testing-library/react";
import { matchers } from "jest-emotion";

expect.extend(matchers);

import wrapWithReduxAndStyle from "../testHelpers/wrapWithReduxAndStyle";
import AddPlayerTurn from "../AddPlayerTurn";
import Player from "../../../logic/Player";
import Planet from "../../../logic/Planet";
import { Simulate } from "react-dom/test-utils";

const testPlayer = new Player("test");
const originPlanet = new Planet("A", testPlayer, { x: 0, y: 0 });
const destinationPlanet = new Planet("B", null, { x: 1, y: 1 });

const amountChange = jest.fn();
const addOrder = jest.fn();
const cancelOrder = jest.fn();

describe("<AddPlayerTurn />", (): void => {
  it("renders <AddPlayerTurn /> component without any selected planets", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <AddPlayerTurn activePlayer={testPlayer} amount={10} onOrderAmountChange={amountChange} onAddOrder={addOrder} onCancel={cancelOrder} />
      )
    );
    expect(container).toMatchSnapshot();
  });

  it("renders <AddPlayerTurn /> component with origin planet selected", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <AddPlayerTurn
          activePlayer={testPlayer}
          originPlanet={originPlanet}
          amount={10}
          onOrderAmountChange={amountChange}
          onAddOrder={addOrder}
          onCancel={cancelOrder}
        />
      )
    );
    expect(container).toMatchSnapshot();
  });

  it("renders <AddPlayerTurn /> component with origin and destination planet selected", (): void => {
    const { container } = render(
      wrapWithReduxAndStyle(
        <AddPlayerTurn
          activePlayer={testPlayer}
          originPlanet={originPlanet}
          destinationPlanet={destinationPlanet}
          amount={10}
          onOrderAmountChange={amountChange}
          onAddOrder={addOrder}
          onCancel={cancelOrder}
        />
      )
    );
    expect(container).toMatchSnapshot();
  });

  it("renders <AddPlayerTurn /> could change amount", (): void => {
    const { getByTestId, getByText } = render(
      wrapWithReduxAndStyle(
        <AddPlayerTurn
          activePlayer={testPlayer}
          originPlanet={originPlanet}
          destinationPlanet={destinationPlanet}
          amount={10}
          onOrderAmountChange={amountChange}
          onAddOrder={addOrder}
          onCancel={cancelOrder}
        />
      )
    );
    // first - check if we are passing event to amount change
    const input = getByTestId("order-amount") as HTMLInputElement;
    const changeEvent = { target: { value: "0" } };
    // seems like issue with types
    Simulate.change(input, changeEvent as any);
    expect(amountChange.mock.calls.length).toBe(1);
    expect(amountChange.mock.calls[0][0]).toMatchObject(changeEvent);
    // next check if we are calling cancel order
    Simulate.click(getByText("Cancel order"));
    expect(cancelOrder.mock.calls.length).toBe(1);
    // now make sure we are able to call create
    Simulate.click(getByText("Add order"));
    expect(addOrder.mock.calls.length).toBe(1);
  });
});
