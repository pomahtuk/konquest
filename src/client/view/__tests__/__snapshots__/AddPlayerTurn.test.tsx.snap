// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddPlayerTurn /> renders <AddPlayerTurn /> component with origin and destination planet selected 1`] = `
.emotion-7 {
  position: relative;
  display: inline-block;
  top: 0;
  left: 0;
}

.emotion-2 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 0 0 1px;
  top: -1px;
  left: -1px;
}

.emotion-3 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 1px 0 0;
  top: -1px;
  right: -1px;
}

.emotion-4 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 0 1px 1px;
  bottom: -1px;
  left: -1px;
}

.emotion-5 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 1px 1px 0;
  bottom: -1px;
  right: -1px;
}

.emotion-6 {
  border: 1px solid #acf9fb;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  vertical-align: middle;
  position: relative;
  background-color: rgba(6,43,45,0.65);
  color: #acf9fb;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  font-family: Arial,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  z-index: 2;
}

.emotion-6:focus {
  outline: none;
}

.emotion-1 {
  border: none;
  padding: 0;
  margin: 0;
}

.emotion-0 {
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  background: #fff;
  border-radius: 2px;
  border: 1px solid #E6E6E6;
  display: block;
  width: 100%;
  text-align: left;
  margin: 8px 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  padding: calc(8px + 1px);
  outline: none;
}

.emotion-0:focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(0,113,194,0.3);
  border-color: #0071C2;
  border-radius: 2px;
}

<div>
  <div>
    <h2>
      test
       turn
    </h2>
    <legend>
      Sending fleet from 
      A
       to 
      B
      . Select amount:
    </legend>
    <div
      class="emotion-1"
    >
      <input
        class="emotion-0"
        data-testid="order-amount"
        type="number"
        value="10"
      />
    </div>
    <div
      class="emotion-7"
    >
      <div
        class="emotion-2"
      />
      <div
        class="emotion-3"
      />
      <div
        class="emotion-4"
      />
      <div
        class="emotion-5"
      />
      <button
        class="emotion-6"
        type="button"
      >
        Add order
      </button>
    </div>
    <div
      class="emotion-7"
    >
      <div
        class="emotion-2"
      />
      <div
        class="emotion-3"
      />
      <div
        class="emotion-4"
      />
      <div
        class="emotion-5"
      />
      <button
        class="emotion-6"
        type="button"
      >
        Cancel order
      </button>
    </div>
  </div>
</div>
`;

exports[`<AddPlayerTurn /> renders <AddPlayerTurn /> component with origin planet selected 1`] = `
.emotion-5 {
  position: relative;
  display: inline-block;
  top: 0;
  left: 0;
}

.emotion-0 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 0 0 1px;
  top: -1px;
  left: -1px;
}

.emotion-1 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 1px 1px 0 0;
  top: -1px;
  right: -1px;
}

.emotion-2 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 0 1px 1px;
  bottom: -1px;
  left: -1px;
}

.emotion-3 {
  border-color: #acf9fb;
  z-index: 1;
  opacity: 1;
  position: absolute;
  border-style: solid;
  width: 8px;
  height: 8px;
  border-width: 0 1px 1px 0;
  bottom: -1px;
  right: -1px;
}

.emotion-4 {
  border: 1px solid #acf9fb;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 8px 16px;
  cursor: pointer;
  font-family: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-sizing: border-box;
  vertical-align: middle;
  position: relative;
  background-color: rgba(6,43,45,0.65);
  color: #acf9fb;
  font-size: 14px;
  font-weight: 400;
  line-height: 20px;
  font-family: Arial,sans-serif;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  position: relative;
  z-index: 2;
}

.emotion-4:focus {
  outline: none;
}

<div>
  <div>
    <h2>
      test
       turn
    </h2>
    <div>
      Please select destination planet
    </div>
    <div
      class="emotion-5"
    >
      <div
        class="emotion-0"
      />
      <div
        class="emotion-1"
      />
      <div
        class="emotion-2"
      />
      <div
        class="emotion-3"
      />
      <button
        class="emotion-4"
        type="button"
      >
        Cancel order
      </button>
    </div>
  </div>
</div>
`;

exports[`<AddPlayerTurn /> renders <AddPlayerTurn /> component without any selected planets 1`] = `
<div>
  <div>
    <h2>
      test
       turn
    </h2>
    <div>
      Please select origin planet
    </div>
  </div>
</div>
`;
